feat(test): add `URL` field to annotations for hyperlink display

osohyun0224 opened the pull request:
related issue: https://github.com/microsoft/playwright/issues/30095

**Description:**
This pull request introduces the ability to include `URL` directly within annotations in Playwright tests, enabling the URLs to be displayed as hyperlinks in test reports.

**Motivation:**
Current annotation functionality only allows for text descriptions, which limits the ability to navigate directly to related resources or issue trackers. By integrating a URL field, we provide a more interactive and informative testing environment.

**Changes:**
Added an optional url field to the Annotation type to accommodate `URL`
Updated type definitions and documentation to reflect the addition.

**Expected Outcome:**
With the addition of the url field, annotations can now directly link to URLs, which are displayed as hyperlinks in test reports. This update improves the functionality of test descriptions and enhances the accessibility of related online resources.

**References:**
Closes https://github.com/microsoft/playwright/issues/30095 - "[Feature]: Show URL's within annotations as hyperlink"

**Code Check**

<details>
<summary>toggle dropdown</summary>
<div markdown="1">

I have written a test to ensure that the changes were applied correctly, and I have verified its functionality through screenshots. I examined whether the test is running as expected by capturing images of the test execution results.

```
it('should not hang on ctrl-click during provisional load', async ({ context, page, server, isWindows, browserName, isLinux }) => {
  it.info().annotations.push({ type: 'issue', url: 'https://github.com/microsoft/playwright/issues/11595' });
....
```

![image](https://github.com/microsoft/playwright/assets/53892427/340a4da5-c7da-45d4-8f72-91161f521a71)
</div>
</details>

osohyun0224 committed:
feat(test): Add 'url' field to annotations

The 'url' field is added to the test annotations to provide a distinct way to store links alongside descriptions. This change allows annotations to include URLs for improved clarity and direct linking, which enhances test report organization.

References: #30095

osohyun0224 committed:
Merge branch 'microsoft:main' into feature/annotation-hyperlinks

osohyun0224 commented:
@microsoft-github-policy-service agree

injae-kim commented on packages/playwright/types/test.d.ts:
```diff
@@ -8120,6 +8120,11 @@ export interface TestInfo {
      */
     description?: string;
+
+    /**
+      * Optional URL to provide additional context or link to an issue tracker.
+    */
```
```suggestion
    /**
     * Optional URL to provide additional context or link to an issue tracker.
     */
```

osohyun0224 committed:
fix: formatting explanation code

osohyun0224 committed:
feat(types): add 'url' field to TestCaseAnnotation type

Added an optional 'url' field to the TestCaseAnnotation type to support hyperlink functionality within test annotations. This update allows annotations to include URLs, enhancing their utility by providing direct links to related resources or issue trackers.

osohyun0224 committed:
fix(docs): fix a typo

osohyun0224 committed:
feat(test): Add URL field to the basic test code

pavelfeldman commented:
This change looks good. Could you follow up with:
- rendering it in the HTML report and other reports
- rendering it in the Trace
- supporting it in the TeleReceiver/Emitter for shard merger?

saurabhdes commented:
thanks @osohyun0224 for your change, will this change also fix https://github.com/microsoft/playwright/issues/30140 ? or that's a separate issue

osohyun0224 commented:
@saurabhdes Hello :) I checked your issue, and I think it's an issue related to the newly added feature. I kept the title and description as is and added a new url field. But the functional purpose of your issue is the same ;) I am planning to develop additional features suggested by the maintainer.
